p2panda-rs/src/next/operation/operation.rs

```diff
@@ -0,0 +1,364 @@
+        self.previous_operations = Some(previous_operations.to_owned());
+        self
+    }
+
+    /// Set operation fields.
+    pub fn fields(mut self, fields: &[(&str, OperationValue)]) -> Self {
+        let mut operation_fields = OperationFields::new();
+
+        for (field_name, field_value) in fields {
+            if operation_fields
+                .insert(field_name, field_value.to_owned())
+                .is_err()
+            {
+                // Silently fail here as the underlying data type already takes care of duplicates
+                // for us ..
+            }
+        }
+
+        self.fields = Some(operation_fields);
+        self
+    }
+
+    /// Builds and returns a new `Operation` instance.
+    ///
+    /// This method checks if the given previous operations and operation fields are matching the
+    /// regarding operation action.
+    pub fn build(&self) -> Result<Operation, OperationBuilderError> {
+        let operation = Operation {
+            action: self.action,
+            version: OperationVersion::V1,
+            schema: self.schema.to_owned(),
+            previous_operations: self.previous_operations.to_owned(),
+            fields: self.fields.to_owned(),
+        };
+
+        validate_operation(&operation, &self.schema)?;
+
+        Ok(operation)
+    }
+}
+
+/// Operations describe data mutations of "documents" in the p2panda network. Authors send
+/// operations to CREATE, UPDATE or DELETE documents.
+///
+/// The data itself lives in the "fields" object and is formed after an operation schema.
+///
+/// Starting from an initial CREATE operation, the following collection of UPDATE operations build
+/// up a causal graph of mutations which can be resolved into a single object during a
+/// "materialisation" process. If a DELETE operation is published it signals the deletion of the
+/// entire graph and no more UPDATE operations should be published.
+///
+/// All UPDATE and DELETE operations have a `previous_operations` field which contains a vector of
+/// operation ids which identify the known branch tips at the time of publication. These allow us
+/// to build the graph and retain knowledge of the graph state at the time the specific operation
+/// was published.
+#[derive(Clone, Debug, PartialEq)]
+pub struct Operation {
+    /// Version of this operation.
+    pub(crate) version: OperationVersion,
+
+    /// Describes if this operation creates, updates or deletes data.
+    pub(crate) action: OperationAction,
+
+    /// Schema matching this operation.
+    pub(crate) schema: Schema,
+
+    /// Optional document view id containing the operation ids directly preceding this one in the
+    /// document.
+    pub(crate) previous_operations: Option<DocumentViewId>,
+
+    /// Optional fields map holding the operation data.
+    pub(crate) fields: Option<OperationFields>,
+}
+
+impl AsOperation for Operation {
```

Comment:
Did you have a "next step" in mind here? I see that the implementations of `Schematic` and `Actionable` duplicate the methods implemented in `AsOperation`. Should we favour the new traits and remove the old one?

Reply:
Hmm... I see the `fields()` in `Actionable` return `PlainFields` maybe this is important...

Reply:
They are close but still for different purposes, `AsOperation` was written with storage in mind. I think it's fine to leave it as is even when its partly redundant, otherwise we might end up with too strong ties in the next refactoring when suddenly storage needs new things, no shame in ignoring DRY 😅✌️

At least not after we learned a bit about the API?

Reply:
Yeh, I like how traits can explain the properties of a struct, want to understand the implications of splitting it up like this though. Not just a question of DRYness :desert: ;-p

Reply:
> AsOperation was written with storage in mind

This isn't true actually, `AsOperation` wasn't about storage, it was a way to abstract out the properties of an `Operation`, these are present on `Operation` & `VerifiedOperation`.

Reply:
Ah I see! Yeag, if you have a suggestion to bring stuff together please go ahead 👍

Reply:
If you're interested in the design / background thinking, this might help: https://github.com/p2panda/handbook/issues/187

Reply:
Thanks, that's good background.

I see the need for the traits in the new validation patterns, but the existing `AsOperation` trait is performing the same function (abstracting the same methods). All I'd suggest is removing one or the other. So `PlainOperation`, `Operation` and `VerifiedOperation` all implement the same trait(s) related to being an operation. 

The more specific naming `Actionable` and `Schematic` are cool how they define the specific functionality they provide. My instinct is that it's a little too specific though, as these properties are an inherent part of operations themselves, we can't have an un-actionable or un-schematic operation (i don't think...?), so I'm not sure the benefit of splitting them up is (although I actually like it in some way, but maybe that's just cos I'm so deep in it though...).

Reply:
> The more specific naming Actionable and Schematic are cool how they define the specific functionality they provide. My instinct is that it's a little too specific though, as these properties are an inherent part of operations themselves, we can't have an un-actionable or un-schematic operation (i don't think...?), so I'm not sure the benefit of splitting them up is (although I actually like it in some way, but maybe that's just cos I'm so deep in it though...).

Currently that's (kinda) true and one could see the new traits as something too far in the future? I don't really mind going in either direction, the difference is maybe that the new traits open more doors, but again, I don't know if their design really reflects our protocol yet (we didn't reflect upon what operations are yet):

* DELETE operations are currently only half "schematic", they contain a schema id but nothing to validate against (fields), I think there is a hole in our current thinking
* We probably want non-operations to implement some sub features of operations, like `DocumentView` to implement `Schematic`, as we need to validate materialized views against schemas after decrypting their fields
* Not sure about some future system schemas, they might not need a schema id at all?

You can see, all of these are a little too far in the future, I'm happy to look into this later or keep both worlds as it us in this PR right now to explore that direction more.

Reply:
Designing traits is special I realize and learning right now: you _can_ reflect more abstract (like protocol-level) designs with it but also want to offer some pragmatic API guarantees (like structs which have the same fields, even though they are different).

Reply:
The `AsOperation` concept can maybe be described as: we have `Operation`, `VerifiedOperation` and `PlainOperation` and all have an action and version field and theoretically can be validated against a schema etc.

Reply:
The "new" direction somewhat hints on having distinct structs for each action, ala `DeleteOperation` etc. instead of a single Operation struct which accommodates everything. The more different new features (Key group, secret group, blobs, ...) we introduce the clearer we might need to separate them with types and traits. That's future thinking again but it was tempting to go there during implementation of this PR.

Reply:
Your todo commit decided for AsOperation I can see 😅

Reply:
Thanks for all the explanations here.

I see that these traits are a step towards some future coolness, and even though there are some possibilities for refactoring them to work more nicely with the existing operation traits (just sugar really), it's fine with me to just leave them all existing as they are now.